hail/src/main/scala/is/hail/rvd/AbstractRVDSpec.scala

```diff
@@ -233,8 +233,16 @@ case class IndexSpec2(_relPath: String,
 
 object IndexSpec {
   def fromKeyAndValuePTypes(relPath: String, keyPType: PType, annotationPType: PType, offsetFieldName: Option[String]): AbstractIndexSpec = {
-    val leafNodeSpec = TypedCodecSpec(LeafNodeBuilder.typ(keyPType, annotationPType), BufferSpec.default)
-    val internalNodeSpec = TypedCodecSpec(InternalNodeBuilder.typ(keyPType, annotationPType), BufferSpec.default)
+    val leafType = LeafNodeBuilder.typ(keyPType, annotationPType)
+    val leafNodeSpec = TypedCodecSpec(
+      EType.defaultFromPType(leafType, useTransposedArrayOfStructs = false),
+      leafType.virtualType,
+      BufferSpec.default)
+    val internalType = InternalNodeBuilder.typ(keyPType, annotationPType)
```

Comment:
Does this break when you don't have the parameterization?

Reply:
If so, we should probably fix it instead of adding the parameterization.

Reply:
Yes. See https://github.com/hail-is/hail/pull/7803/commits/dcab64c108a3f647480dc104c2c6ef00f2f0c98d

I'd be very happy to fix the types around indices, and fix their specs/metadata.

Reply:
OK, great. I'll refresh myself on how they're broken.

Reply:
Actually. I think that we did a pretty good job here for things in the wild. The tests are where the issue seems to be.

Reply:
okay, I think I understand the problems with bgens now -- can we back out these changes? I think these are just copying the implementation of `TypedCodecSpec.apply(ptype, bufferspec)`

Reply:
And I don't think bgens should really be causing a huge amount of trouble for the time being, since we're hardcoding the encoded type of the index files.